api/handlers/listhandler.py

```diff
@@ -275,6 +283,55 @@ def put(self, cont_name, list_name, **kwargs):
             return {'modified':result.modified_count}
 
 
+class TagsListHandler(ListHandler):
+    """
+    TagsListHandler overrides put, delete methods of ListHandler to propagate changes to group tags
+    If a tag is renamed or deleted at the group level, project, session and acquisition tags will also be renamed/deleted
+    """
+
+    def put(self, cont_name, list_name, **kwargs):
+        _id = kwargs.get('cid')
+        result = super(TagsListHandler, self).put(cont_name, list_name, **kwargs)
+        if cont_name == 'groups':
+            payload = self.request.json_body
+            current_value = kwargs.get('value')
+            new_value = payload.get('value')
+            query = {'$and':[{'tags': current_value}, {'tags': {'$ne': new_value}}]}
+            update = {'$set': {'tags.$': new_value}}
+            self._propagate_group_tags(_id, query, update)
+        return result
+
+    def delete(self, cont_name, list_name, **kwargs):
+        _id = kwargs.get('cid')
+        result = super(TagsListHandler, self).delete(cont_name, list_name, **kwargs)
+        if cont_name == 'groups':
+            deleted_tag = kwargs.get('value')
+            query = {}
+            update = {'$pull': {'tags': deleted_tag}}
+            self._propagate_group_tags(_id, query, update)
+
+    def _propagate_group_tags(self, _id, query, update):
```

Comment:
db operations should go in the `dao` module or in a submodule. We made some exceptions in the past, but it seems the right thing to do here (maybe together with `_propagate_project_permissions` in the permission handler).